src/main/webapp/src/sessions/gapi.session.ts

```diff
@@ -0,0 +1,143 @@
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {NgZone, Injectable} from '@angular/core';
+import {environment} from '../environments/environment';
+import {TaskComponent} from '../app/calendar-task/task.component';
+import {CLIENT_ID} from '../app/SensitiveData';
+
+const GAPI_CLIENT_ID = CLIENT_ID;
+const API_KEY = environment.calendar_key;
+
+const DISCOVERY_DOCS = [
+  'https://www.googleapis.com/discovery/v1/apis/drive/v3/rest',
+];
+const SCOPES = 'https://www.googleapis.com/auth/calendar';
+
+declare const gapi: any;
+
+@Injectable()
+export class GapiSession {
+  public auth2: any;
+  hasConsent = false;
+  tasks = new TaskComponent();
+
+  constructor(private zone: NgZone) {}
+
+  ngOnInit() {
+    this.loadClient();
+  }
+
+  /**
+   * Load the Calendar API client and provide authentication
+   */
+  loadClient(): void {
+    gapi.load('client:auth2', () => {
+      gapi.auth2
+        .init({
+          apiKey: API_KEY,
+          clientId: GAPI_CLIENT_ID,
+          discoveryDocs: DISCOVERY_DOCS,
+          scope: SCOPES,
+        })
+        .then(auth => {
+          this.zone.run(() => {
+            this.auth2 = auth;
+          });
+        });
+    });
+  }
+
+  /**
+   * Returns whether user is signed in with Google
+   */
+  get isSignedIn(): boolean {
+    return this.auth2.isSignedIn.get();
+  }
+
+  /**
+   * Returns whether the user has accepted the consent prompt in the current session
+   */
+  get consent(): boolean {
+    return this.hasConsent;
+  }
+
+  /**
+   * Google Authentication; displays consent prompt for Google Calendar API
+   */
+  signIn() {
+    return this.auth2
+      .signIn({
+        prompt: 'consent',
+      })
+      .then(() => {
+        this.hasConsent = true;
+      });
+  }
+
+  /**
+   * Sign out from Google Auth
+   */
+  signOut(): void {
+    this.auth2.signOut();
+  }
+
+  /**
+   * Retrieve calendar events given a date and call function to display the events in the page
+   * @param selectedDate - start date and time (min)
+   * @param selectedDateMax - end of the day of selected date (max)
+   */
+  listEvents(selectedDate: string, selectedDateMax: string) {
+    gapi.client.load('calendar', 'v3', () => {
+      gapi.client.calendar.events
+        .list({
+          calendarId: 'primary',
+          timeMin: selectedDate,
+          timeMax: selectedDateMax,
+          singleEvents: true,
+          orderBy: 'startTime',
+        })
+        .then(response => {
+          const events = response.result.items;
+          this.tasks.clearTasks();
+          if (events.length > 0) {
+            for (let i = 0; i < events.length; i++) {
```

Comment:
I think I missed this but you should also be able to use [`forEach`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach)